### ISSUE TITLE
patterns: catch zero repeats

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#3002]
Patterns: in some situations, a zero repeats can hang sclang. This is known, but arguably it could be avoided in many cases.

Here is one reproducer:

```
a = 0; // imagine this be a varying variable
b = Pn(Pseries(0, 1, a));
x = a.asStream;
x.next; // hangs immediately
```

A possible fix is in Pseries:

```
Pseries : Pattern {	// arithmetic series
	var <>start=0, <>step=1, <>length=inf;
	*new { arg start = 0, step = 1, length=inf;
		^super.newCopyArgs(start, step, length)
	}
	storeArgs { ^[start,step,length] }

	embedInStream { arg inval;
		var outval, counter = 0;
		var cur = start.value(inval);
		var len = length.value(inval);
		var stepStr = step.asStream, stepVal;
		if(len <= 0) { nil.alwaysYield }; // <-------------- just check for boundary condition
		while { counter < len } {
			stepVal = stepStr.next(inval);
			if(stepVal.isNil) { ^inval };
			outval = cur;
			cur = cur + stepVal;
			counter = counter + 1;
			inval = outval.yield;
		};
		^inval;
	}
}
```

I've made a branch here, compare: https://github.com/supercollider/supercollider/compare/master...telephon:topic-catch-pattern-length-0

I can see if there are more and similar cases.

**Comment by telephon** [comment]
Here is a list of all patterns I found which have `length` or `repeats` in their `*new` message:

```

Pseries
Pcbrown
Pgeom
Pprob
Pbrown
PintC
PintN
Ppoisson
Pcauchy
PintL
Pbeta
Pgauss
Pwhite
Pbrown
Pwhite
Pwhite
Pexprand
Pwhite


Pstep
Pboolnet
Pindex
Pget
Pkey
Pdict
Peventmod
Pfsm2
Ptime
Peuclid
Pfuncn
Ppatmod
Pseg
Pspy
Pn
Pgate
Pn
Pmswitch
ListPattern
Pwrand
Pdfsm
Pabsolute
Pseq
Pslide
Phrand
ListPattern
Pabsolute

```

```
Pattern.allSubclasses.collect { |class| class.class.findRespondingMethodFor(\new) }.select { |meth| meth.argNames.includes(\length) }.do { |meth| meth.ownerClass.postln }
```

We have to see if it is really a good idea to yield nil instead of just returning `^inval`

**Comment by jamshark70** [comment]
```
(
// skip every 4th pattern
var numStream = Pseq([5, 4, 3, 0], inf).asStream;

p = Pbind(
	// 3 items vs 4 'num' values = 12-phrase cycle
	\degree, Pseq([
		Pseries({ rrand(-2, 2) }, 1, numStream),
		Pseries({ rrand(5, 9) }, -1, numStream),
		Pseries({ rrand(-9, -5) }, 3, numStream)
	], inf),
	\dur, 0.125
).play;
)

p.stop;
```

Zero repeats may be compositionally useful to suppress sub-streams from an outer-layer Pseq.

**Comment by jamshark70** [comment]
Here is an alternate approach that I use for patterns played on a clock: https://github.com/jamshark70/ddwPatterns/blob/master/ddwFilterPatterns.sc#L35

It could be adapted for Pn -- say, if Pn detects that an iteration embedded no items, it could terminate rather than looping. How to know that is perhaps tricky (which is maybe why I used the clock).

**Comment by telephon** [comment]
Ah interesting, yes. Probably this should even be a `loop` method outside the pattern system. If we could somehow inscribe a call id in a context, this might be possible.

`Maybe` catches recursions that way: https://github.com/supercollider/supercollider/blob/master/SCClassLibrary/JITLib/Patterns/Fdef.sc#L141

I don't have an immediate idea how to use a similar way for functions.

**Comment by telephon** [comment]
replaced by #3603

**Comment by brianlheim** [comment]
(#3603 hasn't been merged yet)

**Comment by jamshark70** [comment]
IMO this needs more discussion.

#3603 contains three fixes. Two of them are redundant and, I think, should be removed. I commented about it over there.

The third is properly considered an optimization: If a pattern initializes some heavy resources (`Pgpar` makes some groups), but a given instance also specifies 0 repeats, you can skip the initialization.

So #3603 does not address the initial issue reported here (infinite looping).

The problem with infinite looping is not the zero-length pattern itself. The infinite loop happens only if the zero-length pattern is embedded in a higher-level pattern with infinite repeats. (Note that the reproducer is not only `Pseries` -- it's `Pn(Pseries(...), inf)` -- without the `Pn`, you don't have the problem.)

The higher-level pattern, currently, has no way to know if its subpattern(s) yielded anything or not -- so, it blindly repeats, unaware that nothing in the subpattern will pause the thread.

Therefore, the only logically correct solution is to make the higher-level looping pattern aware of this.

It's tricky because we already depend on the return value from `embedInStream` for the next inval. Returning an array (`#inval, didYield = something.embedInStream(inval)`) is a risky API change that's guaranteed to break something -- bad idea.

Possibly less risky: A pattern that embeds something else could pass in a flag by reference. Child patterns would modify the flag.

```
Pn : FilterPattern {
	var <>repeats, <>key;
	*new { arg pattern, repeats=inf, key;
		^super.newCopyArgs(pattern, repeats, key )
	}
	storeArgs { ^[pattern,repeats, key] }
	embedInStream { | event |
		if(key.isNil) {
			repeats.value(event).do {
				var didYield = Ref(false);
				event = pattern.embedInStream(event, didYield);
				if(didYield.value.not) { ^event };
			};
		} {
			repeats.value(event).do {
				var didYield = Ref(false);
				event[key] = true;
				event = pattern.embedInStream(event, didYield);
				if(didYield.value.not) { ^event };
			};
			event[key] = false;
		};
		^event;
	}
}

Pseries : Pattern {	// arithmetic series
	var <>start=0, <>step=1, <>length=inf;
	*new { arg start = 0, step = 1, length=inf;
		^super.newCopyArgs(start, step, length)
	}
	storeArgs { ^[start,step,length] }

	embedInStream { arg inval, didYield(Ref(false));
		var outval, counter = 0;
		var cur = start.value(inval);
		var len = length.value(inval);
		var stepStr = step.asStream, stepVal;
		while { counter < len } {
			stepVal = stepStr.next(inval);
			if(stepVal.isNil) { ^inval };
			outval = cur;
			cur = cur + stepVal;
			counter = counter + 1;
			didYield.value = true;
			inval = outval.yield;
		};
		^inval;
	}
}
```

But that would require changes to every pattern (including user extension patterns). It "works" but getting everyone to update is a mess.

Or, add a variable to stream classes for the number of items yielded (which `yield` would update). `Pn` could then save `var numBefore = thisThread.numYielded`, and after embedding, if `thisThread.numYielded == numBefore`, then you know nothing happened and it's time to quit.

I also like the idea from my `PnNilSafe` to have a maximum number of zero-yield iterations. It *is* valid, if controlled, for a subpattern to do nothing. Allowing some number of empty iterations before quitting means that the system is tolerant of valid uses, but won't hang in case of accidents.

Let's say, for instance, I want to play scale segments, where each segment starts on successive scale degrees, and I want to skip 35% of starting scale degrees. You can skip a starting scale degree by playing a zero-length segment for it.

```
(
var start = Pseries(-14, 1, inf).asStream,
len = Pswitch1([
	0,
	Pwhite(3, 6, inf)
], Pwrand([0, 1], [0.35, 0.65], inf)).asStream;

p = Pbind(
	\degree, Pn(Pseries(start, 1, len), inf),
	\dur, 0.125
).play;
)

p.stop;
```

Any solution to the infinite loop problem should not rule out this case.

**Comment by jamshark70** [comment]
Also the unit test in the PR is useful -- worth adding, but not enough to close this issue.

**Comment by telephon** [comment]
Yes, that is true. The fix in  #3002 is for something else.

**Comment by telephon** [comment]
There is another way in which we could protect from infinite loops. I'm not sure if it'll be the best way to do it, but just to continue thinking this through.

It would involve two things: 

1. any stream (below it is `x`)  that has a potentially infinite repetition could yield a special value (here called `\blocking`) as a **last value**. This value should be essentially a `nil`, but which doesn't terminate a stream.
2. The outermost player routine is structured so that it terminates any stream that starts with a `blocking` value. Otherwise, it filters out the `\blocking`values and passes on all stream values.

```supercollider

// this is our basic situation
c = 0.5.coin;
f = Pfunc { if(c) { 1.yield } }.asStream;
x = Routine { loop { f.embedInStream } };


// this could be done about it:
// in a loop where we don't know for sure if it will 
(
a = { |fails = true|
	p = { |r|
		Routine {
			var first;
			first = r.next;
			if(first != \blocking) {
				first.yield;
				r.select(_ != \blocking).embedInStream
			} {
				nil.alwaysYield
			}
		}
	};
	
	f = Pfunc { if(fails.not) { 1.yield } }.asStream;
	x = Routine { loop { f.embedInStream; \blocking.yield } };
	p.(x)
};
)

y = a.(fails: true); // -> [ nil, nil, nil, nil, nil, nil, nil, nil ]
y.nextN(8);


y = a.(fails: false); // -> [ 1, 1, 1, 1, 1, 1, 1, 1 ]
y.nextN(8);


```
